### PR TITLE
add type annotations to common_nn.py

### LINKED ISSUE #48189: Enable torch.testing._internal.common_nn.* typechecks during CI
`TestTypeHints.test_run_mypy` should pass with `[mypy-torch.testing._internal.common_nn.*]` deleted from mypy.ini

cc @ezyang @malfet @rgommers @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Fixes #48189

### CODE REVIEW

File: torch/testing/_internal/common_nn.py [hunk 1]
```diff
@@ -640,7 +643,7 @@ def nllloss_no_reduce_test():
     return dict(
         fullname='NLLLoss_no_reduce',
         constructor=wrap_functional(
-            lambda i: F.nll_loss(i, t.type_as(i).long(), **kwargs)),
+            lambda i: F.nll_loss(i, t.type_as(i).long(), **kwargs)),  # type: ignore[arg-type]
```
Comment (rgommers) [inline torch/testing/_internal/common_nn.py:646]: Why add the ignores for `**kwargs`, for readability of the test? I understand not replacing it with `reduction='none'` (which works fine), but you could replace with:
```
 F.nll_loss(..., reduction=kwargs['reduction'])
```
Then the test remains understandable (the `kwargs` above the `return` is what varies between tests), but the ignores aren't needed.

The advantage of that is that it adds to the `nll_loss` testing with type annotations (plus, the fewer ignores, the better).

File: torch/testing/_internal/common_nn.py [hunk 2]
```diff
@@ -3691,7 +3694,7 @@ def kldivloss_reference(input, target, reduction='mean'):
         return result.mean()
     elif reduction == 'sum':
         return result.sum()
-    elif reduction == 'batchmean' and results.dim() != 0:
+    elif reduction == 'batchmean' and result.dim() != 0:
```
Comment (rgommers) [inline torch/testing/_internal/common_nn.py:3697]: This is clearly unused code (because the variable doesn't exist), not sure if that's intended or not. @nikitaved you added this in commit 35cdb785228b8, can you comment on how this `_reference` stuff was used?

File: torch/testing/_internal/common_nn.py [hunk 3]
```diff
@@ -4631,6 +4634,26 @@ def padding3d_circular(input, pad):
 
 class NNTestCase(TestCase):
 
+    # _forward is defined in classes inheriting from TestCase
```
Comment (rgommers) [inline torch/testing/_internal/common_nn.py:4637]: nit: last word of this comment should be `NNTestCase`

File: torch/testing/_internal/common_nn.py (hunk 1 above)
Comment (rgommers) [reply torch/testing/_internal/common_nn.py:646]: This applies to other ignores below as well.

**Review by rgommers — CHANGES_REQUESTED** [review]
Thanks @guilhermeleobas. A couple of things to fix, mainly getting rid of some `#type: ignore`'s that aren't necessary.

File: torch/testing/_internal/common_nn.py (hunk 2 above)
Comment (nikitaved) [reply torch/testing/_internal/common_nn.py:3697]: @rgommers , I am not sure I understand. I did not add any `_reference`, and did not touch any `reduction` options...

Comment (nikitaved) [reply torch/testing/_internal/common_nn.py:3697]: @rgommers , the reference is used in the tests, I did not introduce `kldivloss_reference`, but I did introduce `kldiv_log_target_reference` which was a copy of `kldivloss_reference`.

Comment (nikitaved) [reply torch/testing/_internal/common_nn.py:3697]: yes, my method suffers from the same issue.

File: torch/testing/_internal/common_nn.py (hunk 1 above)
Comment (guilhermeleobas) [reply torch/testing/_internal/common_nn.py]: Done! I'll ping you if/when CI passes

File: torch/testing/_internal/common_nn.py [hunk 4]
```diff
@@ -2026,21 +2029,21 @@ def fractional_max_pool3d_test(test_case):
     ),
     dict(
         fullname='AvgPool2d_divisor',
-        constructor=lambda: nn.AvgPool2d((2, 2), divisor_override=1),
+        constructor=lambda: nn.AvgPool2d((2, 2), divisor_override=1),  # type: ignore[arg-type]
```
Comment (rgommers) [inline torch/testing/_internal/common_nn.py:2032]: This is a bug in the annotation of `AvgPool2d` (divisor_override should be `int` not `bool`) that should be fixed instead of ignoring it here.

File: torch/testing/_internal/common_nn.py [hunk 5]
```diff
@@ -2026,21 +2029,21 @@ def fractional_max_pool3d_test(test_case):
...
         cpp_constructor_args='torch::nn::AvgPool2dOptions({2, 2}).divisor_override(1)',
         input_size=(2, 3, 6, 6),
         check_with_long_tensor=True,
     ),
     dict(
         fullname='AvgPool2d_divisor_stride',
-        constructor=lambda: nn.AvgPool2d((2, 2), (2, 2), divisor_override=1),
+        constructor=lambda: nn.AvgPool2d((2, 2), (2, 2), divisor_override=1),  # type: ignore[arg-type]
```
Comment (rgommers) [inline torch/testing/_internal/common_nn.py:2039]: remove ignore here

File: torch/testing/_internal/common_nn.py [hunk 6]
```diff
@@ -2026,21 +2029,21 @@ def fractional_max_pool3d_test(test_case):
...
         cpp_constructor_args='torch::nn::AvgPool2dOptions({2, 2}).stride({2, 2}).divisor_override(1)',
         input_size=(2, 3, 6, 6),
         check_with_long_tensor=True,
     ),
     dict(
         fullname='AvgPool2d_divisor_stride_pad',
-        constructor=lambda: nn.AvgPool2d((2, 2), (2, 2), (1, 1), divisor_override=1),
+        constructor=lambda: nn.AvgPool2d((2, 2), (2, 2), (1, 1), divisor_override=1),  # type: ignore[arg-type]
```
Comment (rgommers) [inline torch/testing/_internal/common_nn.py:2046]: and here as well

File: torch/testing/_internal/common_nn.py (hunk 2 above)
Comment (rgommers) [reply torch/testing/_internal/common_nn.py:3697]: okay the rest is used, just not the `batchmean` part - let's leave this as is.

File: torch/testing/_internal/common_nn.py [hunk 7]
```diff
@@ -4837,7 +4864,7 @@ def __call__(self, test_case):
             # TODO: do this with in-memory files as soon as torch.save will support it
             with TemporaryFile() as f:
                 test_case._forward(module, input)
-                torch.save(module, f)
+                torch.save(module, f)  # type: ignore[arg-type]
```
Comment (rgommers) [inline torch/testing/_internal/common_nn.py:4867]: Should also not be ignored, instead fix the `torch.save` annotation by adding `IO[bytes]` - see https://stackoverflow.com/questions/59693359/when-to-use-iostr-iobytes-and-textio-binaryio-in-python-type-hinting

File: torch/testing/_internal/common_nn.py [hunk 8]
```diff
@@ -4903,7 +4930,7 @@ def test_cuda(self, test_case):
             raise unittest.SkipTest('Excluded from CUDA tests')
 
         cpu_input = self._get_input()
-        type_map = {'torch.DoubleTensor': torch.cuda.FloatTensor}
+        type_map = {'torch.DoubleTensor': torch.cuda.FloatTensor}  # type: ignore[attr-defined]
```
Comment (rgommers) [inline torch/testing/_internal/common_nn.py:4933]: `DoubleTensor` et al. are missing from type stubs on purpose, so this ignore is fine

File: torch/testing/_internal/common_nn.py [hunk 9]
```diff
@@ -5188,7 +5215,7 @@ def apply_fn(input, target, *params):
         else:
             inputs = input + params + (target,)
 
-            def apply_fn(input1, input2, target, *params):
+            def apply_fn(input1, input2, target, *params):  # type: ignore[misc]
```
Comment (rgommers) [inline torch/testing/_internal/common_nn.py:5218]: yep, this one is best ignored, too much code duplication to make `mypy` happy here

**Review by rgommers — CHANGES_REQUESTED** [review]
A couple more fixes to make instead of ignoring them.

**Review by rgommers — APPROVED** [review]
LGTM now, thanks @guilhermeleobas

File: torch/nn/modules/pooling.py [hunk 10]
```diff
@@ -601,7 +601,7 @@ class AvgPool2d(_AvgPoolNd):
     count_include_pad: bool
 
     def __init__(self, kernel_size: _size_2_t, stride: Optional[_size_2_t] = None, padding: _size_2_t = 0,
-                 ceil_mode: bool = False, count_include_pad: bool = True, divisor_override: bool = None) -> None:
+                 ceil_mode: bool = False, count_include_pad: bool = True, divisor_override: int = None) -> None:
```
Comment (malfet) [inline torch/nn/modules/pooling.py:604]: Hmm, shouldn't it be `Optional[int]`, shouldn't it?

**Review by malfet — APPROVED** [review]

File: torch/nn/modules/pooling.py (hunk 10 above)
Comment (rgommers) [reply torch/nn/modules/pooling.py:604]: yes indeed. @guilhermeleobas can you update?

Comment (guilhermeleobas) [reply torch/nn/modules/pooling.py]: Done!
